autoload/phpactor/project/bufferMatcher.vim

```diff
@@ -0,0 +1,129 @@
+function! phpactor#project#bufferMatcher#assignFileToProject(filename)
+  if !exists('g:phpactorLoaded')
+    return
+  endif
+
+  if exists('b:project')
+    return
+  endif
+  let l:project = g:phpactorBufferMatcher.matchFileToProject(a:filename)
+  if v:null is l:project
+    return
+  endif
+  let b:project = l:project
+  if g:phpactorBufferMatcher.getRepository().hasProject(b:project)
+    return
+  endif
+
+  call g:phpactorBufferMatcher.getRepository().addProject(b:project)
+endfunction
+
+function! phpactor#project#bufferMatcher#create(projectRepository, projectRootMarkers, filesystemRootMarkers, initialCwd) abort
+  let l:filesystemRootMarkers = a:filesystemRootMarkers
+  if index(l:filesystemRootMarkers, '/') < 0
+    call add(l:filesystemRootMarkers, '/')
+  endif
+
+  return {
+        \ 'initialCwd': a:initialCwd,
+        \ 'projectRootMarkers': a:projectRootMarkers,
+        \ 'filesystemRootMarkers': l:filesystemRootMarkers,
+        \ 'repository': a:projectRepository,
+        \ 'getRepository': function('s:getRepository'),
+        \ 'matchFileToProject': function('s:matchFileToProject')
+        \ }
+endfunction
+
+function s:matchFileToProject(file) dict abort
+  let l:project = self.repository.findProjectContainingFile(a:file)
+
+  if type(l:project) == type({})
+    return l:project
+  endif
+
+  " early version to refactor or change
+  " heredoc are available since vim 8.1.1354
+  let l:message = printf(join([
+        \ 'Phpactor [RPC]',
+        \ 'There is no project enabled for "%s" yet.',
+        \ 'Select the way of assign the project root',
+        \ 'If cancelled, "%s" will be used.'
+        \], "\n"), a:file, self.initialCwd)
+
+  " see phpactor#_path
+  let l:initialDirectory = fnamemodify(a:file, ':p:h')
+  let l:rootDirByMarker = s:searchDirectoryUpwardForRootMarkers(
+        \ l:initialDirectory,
+        \ self.projectRootMarkers,
+        \ self.filesystemRootMarkers
+        \ )
+
+  let l:choices  = [
+          \ l:message,
+          \ printf('1. manual (default "%s")', l:initialDirectory),
+          \ printf('2. use initial working directory: "%s"', self.initialCwd)
+          \ ]
+
+  if v:null != l:rootDirByMarker
+    let l:item = '3. autodetected by root marker: '.l:rootDirByMarker
+    call add(l:choices, l:item)
+  endif
+
+  let l:choice = v:null
+  while index([0,1,2,3], l:choice) < 0
+    " bug: neovim assigns 0 without any interaction, so I used input()
+    let l:choice = inputlist(l:choices)
+    " let l:choice = str2nr(input(l:choices[0], join(l:choices[1:], "\n")))
+    " echo join(l:choices)."\n"
+    " let l:choice = str2nr(input('Enter number: '))
+    redraw
+  endwhile
+
+  if l:choice == 1
+    let l:manualPath = input('Enter file path: ', l:initialDirectory, 'file')
+    redraw
+    if ! empty(glob(l:manualPath.'/'))
+      " todo validate path
+      let l:selectedDir = l:manualPath
+    endif
```

Comment:
What about asking until a valid path is given ?
In case the user made a typo and the path is invalid, we ask him again with the value he just gave us as default.
```suggestion
    let l:selectedDir = s:askForAPath(l:initialDirectory)
```
```vim
function! s:askForAPath(defaultValue) abort
  let l:manualPath = input('Enter file path: ', a:defaultValue, 'file')
  redraw

  if !empty(glob(l:manualPath.'/'))
    return l:manualPath
  endif

  " We might be able to echo here in order to let the user know that the path is invalid
  " If it does not work or is not clean enought, add a flag as parameter or the promp directly
  " And add a short message like "Invalid path\n" to preprend to the prompt
  return s:askPath(l:manualPath)
endfunction
```